idaes/core/util/scaling.py

```diff
@@ -1041,3 +1047,165 @@ def set_derivative_factor_from_state(self, deriv, nominal_wrt=1.0):
+
+
+# New functions
+class ScalingFactorExtractionVisitor(EXPR.StreamBasedExpressionVisitor):
+    """
+    Expression walker for collecting scaling factors in an expression.
+
+    Returns a list of scaling factors for each additive term in the expression.
+    """
+
+    def __init__(self, warning: bool = True, exception: bool = False):
+        """
+        Visitor class used to determine scaling factors of an expression. Do not use
+        this class directly.
+
+        Args:
+            warning: bool indicating whether to log a warning when a
+                missing scaling factors is encountered (default=True)
+            exception: bool indicating whether to raise an Exception when a
+                missing scaling factors is encountered (default=False)
+
+        Notes
+        -----
+        This class inherits from the :class:`StreamBasedExpressionVisitor` to implement
+        a walker that returns the pyomo units and pint units corresponding to an
+        expression.
+        There are class attributes (dicts) that map the expression node type to the
+        particular method that should be called to return the units of the node based
+        on the units of its child arguments. This map is used in exitNode.
+        """
+        super().__init__()
+
+        self.warning = warning
+        self.exception = exception
+
+    def _get_scale_factor_for_sum(self, node, child_scale_factors):
+        sf = []
+        for i in child_scale_factors:
+            for j in i:
+                sf.append(j)
+        return sf
+
+    def _get_scale_factor_for_product(self, node, child_scale_factors):
+        assert len(child_scale_factors) == 2
+
+        sf = []
+        for i in child_scale_factors[0]:
+            for j in child_scale_factors[1]:
+                sf.append(i * j)
+        return sf
+
+    def _get_scale_factor_for_division(self, node, child_scale_factors):
+        assert len(child_scale_factors) == 2
+
+        sf = []
+        for i in child_scale_factors[0]:
+            for j in child_scale_factors[1]:
+                sf.append(i / j)
+        return sf
+
+    def _get_scale_factor_for_power(self, node, child_scale_factors):
+        assert len(child_scale_factors) == 2
+
+        # We will create separate scaling factors for all additive terms
+        # Users can decide how to use these later
+        sf = []
+        for i in child_scale_factors[0]:
+            for j in child_scale_factors[1]:
+                sf.append(i**j)
+        return sf
+
+    def _get_scale_factor_single_child(self, node, child_scale_factors):
+        assert len(child_scale_factors) == 1
+        return child_scale_factors[0]
+
+    def _get_scale_factor_for_unary_function(self, node, child_scale_factors):
+        assert len(child_scale_factors) == 1
+        func_name = node.getname()
+        func = getattr(math, func_name)
+        return [func(i) for i in child_scale_factors[0]]
```

Comment:
If I'm reading this correctly, this is saying the scaling factor for f(x), s_f(x), is the function f evaluated at the scaling factor of x, f(s_x). This is implicitly assuming that f(s_x*x) = f(s_x)*f(x), which is tantamount to saying that f(x) = x^a for some power a.

This is not going to be true for exponential functions: exp(10) ~= 22000, exp(1/10) ~= 1.11., logarithms (unless you're doing stat mech, output of logarithms is generally going to be O(1)) or trig functions.

Reply:
No, that is not quite what this is saying. As written, `sf_x` is effectively the expected magnitude of `x` and thus the expected magnitude of `f(x) ~= f(sf_x)`. If we go and apply the variable scaling factor, we get:

```
x_prime = x*sf_x
expr_prime =f(x_prime/sf_x)
```

I.e. the variable scaling factor gets cancelled out in the end, but the magnitude of the function should be `O(f(sf_x))`.

Reply:
> As written, `sf_x` is effectively the expected magnitude of `x`

For what it's worth, I intended `sf_x` to be the _inverse_ of the expected magnitude of `x`, so that `sf_x * x = O(1)`. It's my understanding that's what IDAES scaling factors are, and if they're not that the documentation needs to be changed.

 That doesn't affect the larger point, though. I'm challenging that `f(x) ~= f(sf_x)` (or `f(1/sf_x)` if we used my convention) in the first place. That's true for powers, but not true for many other classes of function. Exponential functions send nearby values to different orders of magnitude. Logarithms take everything to O(1) (except the sort of inconceivably huge numbers you encounter in Stat Mech). Sine and cosine take things to O(1) without qualification, except you might get unlucky where you evaluate it and get zero out instead.

Reply:
That is true, but we have no way of dealing with that - the only information we have is what the user gave us as the variable scaling factor. Note that this tool is intended to be used to fill in information with the best guesses available based on the variable scaling factors - it is not intended to be a magic bullet. Also, user and model developers can (and should) provide their own constraint scaling where possible - this is just to help us fill in the gaps automatically.

I will also note that sin and cos do not always give O(1) - they have a _maximum_ magnitude of O(1) but can be anything smaller than that too.